### PR TITLE
Refactored user administration library to throw exceptions (main)

### PR DESCRIPTION
**Pull request by korydraughn**
All unit tests pass.

Nothing other than the unit tests and the switch user api plugin uses this library so I don't think core tests need to be run for this.

I'm considering changing more components of the library as well.

TODOs:
- [x] Add unit test for remote zone users
- [x] Add unit test for authentication names (e.g. `iadmin aua`)

### CODE REVIEW

File: lib/administration/user/include/irods/user_administration.hpp [hunk 1]
```diff
@@ -7,342 +7,323 @@
...
+        /// \param[in] conn         The communication object.
+        /// \param[in] user         The user to update.
+        /// \param[in] new_password The new password of the user.
         ///
-        /// \throws user_management_error If the string cannot be converted to an enumeration.
+        /// \return An error code.
         ///
-        /// \return A user_type enumeration.
+        /// \since 4.2.8
+        auto set_user_password(rxComm& conn, const user& user, std::string_view new_password) -> void;
```
Comment (korydraughn) [inline lib/administration/user/include/irods/user_administration.hpp:118]: Consider consolidating the following functions into a single function to bring it closer to other library interfaces.
- set_user_password
- set_user_type
- add_user_auth
- remove_user_auth

The new interface would follow the same pattern as the zone administration library's `modify_zone` function.

File: lib/administration/user/include/irods/user_administration.hpp [hunk 2]
```diff
@@ -7,342 +7,323 @@
...
+            user_type user_type = user_type::rodsuser,
+            zone_type zone_type = zone_type::local) -> void;
 
-        /// \brief The generic exception type used by the user administration library.
+        /// \brief Removes a user from the system.
+        ///
+        /// \param[in] conn The communication object.
+        /// \param[in] user The user to remove.
+        ///
+        /// \return An error code.
```
Comment (alanking) [inline lib/administration/user/include/irods/user_administration.hpp:104]: Add `\throws`?

File: lib/administration/user/include/irods/user_administration.hpp [hunk 3]
```diff
@@ -7,342 +7,323 @@
...
-    ///
-    /// \since 4.2.8
-    namespace NAMESPACE_IMPL
-    {
+        /// \brief Adds a new group to the system.
+        ///
+        /// \param[in] conn  The communication object.
+        /// \param[in] group The group to add.
+        ///
+        /// \return An error code.
```
Comment (alanking) [inline lib/administration/user/include/irods/user_administration.hpp:160]: Add `\throws`?

File: lib/administration/user/include/irods/user_administration.hpp [hunk 4]
```diff
@@ -7,342 +7,323 @@
...
+        ///
+        /// \since 4.2.8
+        auto add_group(rxComm& conn, const group& group) -> void;
+
+        /// \brief Removes a group from the system.
+        ///
+        /// \param[in] conn  The communication object.
+        /// \param[in] group The group to remove.
+        ///
+        /// \return An error code.
```
Comment (alanking) [inline lib/administration/user/include/irods/user_administration.hpp:170]: Add `\throws`?

File: lib/administration/user/include/irods/user_administration.hpp [hunk 5]
```diff
@@ -7,342 +7,323 @@
...
+        /// \since 4.2.8
+        auto remove_group(rxComm& conn, const group& group) -> void;
+
+        /// \brief Adds a user to a group.
+        ///
+        /// \param[in] conn  The communication object.
+        /// \param[in] group The group that will hold the user.
+        /// \param[in] user  The user to add to the group.
+        ///
+        /// \return An error code.
```
Comment (alanking) [inline lib/administration/user/include/irods/user_administration.hpp:181]: Add `\throws`?

File: lib/administration/user/include/irods/user_administration.hpp [hunk 6]
```diff
@@ -7,342 +7,323 @@
...
-        } // namespace v1
+        auto add_user_to_group(rxComm& conn, const group& group, const user& user) -> void;
+
+        /// \brief Removes a user from a group.
+        ///
+        /// \param[in] conn  The communication object.
+        /// \param[in] group The group that contains the user.
+        /// \param[in] user  The user to remove from the group.
+        ///
+        /// \return An error code.
```
Comment (alanking) [inline lib/administration/user/include/irods/user_administration.hpp:192]: Add `\throws`?

**Review by alanking — COMMENTED** [review]
So far, so good

File: lib/administration/user/include/irods/user_administration.hpp (hunk 1 above)
Comment (korydraughn) [reply lib/administration/user/include/irods/user_administration.hpp]: Done.

File: lib/administration/user/include/irods/user_administration.hpp (hunk 2 above)
Comment (korydraughn) [reply lib/administration/user/include/irods/user_administration.hpp]: Done.

File: lib/administration/user/include/irods/user_administration.hpp (hunk 3 above)
Comment (korydraughn) [reply lib/administration/user/include/irods/user_administration.hpp]: Done.

File: lib/administration/user/include/irods/user_administration.hpp (hunk 4 above)
Comment (korydraughn) [reply lib/administration/user/include/irods/user_administration.hpp]: Done.

File: lib/administration/user/include/irods/user_administration.hpp (hunk 5 above)
Comment (korydraughn) [reply lib/administration/user/include/irods/user_administration.hpp]: Done.

File: lib/administration/user/include/irods/user_administration.hpp (hunk 6 above)
Comment (korydraughn) [reply lib/administration/user/include/irods/user_administration.hpp]: Done.

**Comment by korydraughn** [comment]
All unit tests pass.

File: lib/administration/user/src/detail.cpp [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        }
+
+        // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers, readability-magic-numbers)
+        std::array<char, MAX_PASSWORD_LEN + 10> admin_password{};
+
+        // Get the plain text password of the iRODS user.
+        // "obfGetPw" decodes the obfuscated password stored in .irods/irodsA.
+        if (const auto ec = obfGetPw(admin_password.data()); ec != 0) {
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+            THROW(ec, "Failed to unobfuscate password in .irodsA file.");
```
Comment (korydraughn) [inline lib/administration/user/src/detail.cpp:35]: unobfuscate?

Comment (korydraughn) [reply lib/administration/user/src/detail.cpp]: Fixed.

**Comment by korydraughn** [comment]
All unit tests pass.

Any last throughts or concerns before I add the #'s?

**Review by trel — APPROVED** [review]
couldn't find any typos.

looks good to me.

**Review by alanking — APPROVED** [review]

**Comment by korydraughn** [comment]
Added #'s

**Pull request merged**